### DISCUSSION TITLE
Woefully split logging concerns - How do I log the metrics I want?

### DISCUSSION THREAD

**Discussion by nervecenter** [Lightning-AI/pytorch-lightning#7785] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I've been trying to grapple with systems such as `CSVLogger`, `LearningRateMonitor`, and the `training_epoch_end()` and `validation_epoch_end()` methods. Everything seems horribly split, to the point where `LearningRateMonitor` is passed in as a callback, while normal logger objects are passed in to the `Trainer`, and they can't talk to the `LearningRateMonitor`; the passed-in logger objects can't even access the learning rate. Not to mention that manually writing metrics to a logger versus passing logging parameters to a logging object cause them to write mangled and interspersed data to file, rendering the log files a nightmare (or even impossible) to interpret.

All I want to do is write a simple CSV. At the end of each epoch, after validation, I want to record four columns: The epoch number; the average training loss; the average validation loss; the learning rate.

That's it. I've been scouring the docs and it currently seems like it can't be done.

**Comment by Troublem1** [comment]
The learning rate monitor seems to act like a callback so try passing it into the callback hook in the trainer..

**Reply by Troublem1** [reply]
For your objective I propose  you return both the train loss and Val loss at the training step and validation step. Find their averages in Validation at epoch end.
Add the results as a row to your dataframe with its 4 columns..

To get the learning rate at each epoch or step I propose using self.lr so you'd be able get the current learning rate being used at each step/epoch

**Comment by Troublem1** [comment]
Also you don't need to manually do all this ..you can simply use self.log() for both the training and Val averages
